Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch

```diff
@@ -0,0 +1,1123 @@
+From 359544c68aecc33c5f945f827cccec97465f226f Mon Sep 17 00:00:00 2001
+From: egg82 <eggys82@gmail.com>
+Date: Fri, 26 Jul 2019 12:27:13 -0600
+Subject: [PATCH] Add Zstd/LZ4/(GZIP write) support for region files
+
+Zstd using zstd-jni: https://github.com/luben/zstd-jni
+LZ4 using lz4-java: https://github.com/lz4/lz4-java
+There's a new option in PaperConfig for this which defaults to false.
+Since the patch hooks Mojang's versioning system for new compression methods, it
+should be internally consistent, easily-updatable, and reversible with a force-upgrade.
+
+This will likely conflict with plugins and programs expecting region files to be compressed using the current standard.
+
+diff --git a/pom.xml b/pom.xml
+index beda5dc8a..6ef72b759 100644
+--- a/pom.xml
++++ b/pom.xml
+@@ -106,6 +106,17 @@
+             <version>5.1.47</version>
+             <scope>runtime</scope>
+         </dependency>
++        <!-- Paper - Zstd/LZ4 for region files -->
++        <dependency>
++            <groupId>com.github.luben</groupId>
++            <artifactId>zstd-jni</artifactId>
++            <version>1.4.2-1</version>
++        </dependency>
++        <dependency>
++            <groupId>org.lz4</groupId>
++            <artifactId>lz4-java</artifactId>
++            <version>1.6.0</version>
++        </dependency>
+         <!-- testing -->
+         <dependency>
+             <groupId>junit</groupId>
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index 5942c3438..117ffd4ff 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -388,4 +388,32 @@ public class PaperConfig {
+         maxBookPageSize = getInt("settings.book-size.page-max", maxBookPageSize);
+         maxBookTotalSizeMultiplier = getDouble("settings.book-size.total-multiplier", maxBookTotalSizeMultiplier);
+     }
++
++    public static int compressionMethod = 2;
++    private static void compressionMethod() {
++        compressionMethod = getInt("settings.compression-method", 2);
++        if (compressionMethod < 0 || compressionMethod > 127 || (compressionMethod > 2 && compressionMethod < 126)) {
++            fatal("compression-method must be 0, 1, 2, 126, or 127");
++        }
++
++        if (compressionMethod == 127) {
```

Comment:
static final constants for these

Reply:
enum with a string and an int instead. Then we can just enumerate the enum to print out all valid possibilities on the fatal.